pkg/lexer/state.go

```diff
@@ -0,0 +1,326 @@
+// Copyright © 2021 Rak Laptudirm <raklaptudirm@gmail.com>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lexer
+
+import (
+	"fmt"
+	"unicode"
+
+	"github.com/raklaptudirm/mash/pkg/token"
+)
+
+var ErrUnexpectedEOF = fmt.Errorf("unexpected EOF")
+
+func (l *lexer) run() {
+	lexBase(l)
+	close(l.Tokens)
+}
+
+func lexBase(l *lexer) {
+next:
+	r := l.peek()
+	switch {
+	case r == eof:
+		l.emit(token.SEMICOLON)
+		l.emit(token.EOF)
+		return
+	case unicode.IsSpace(r):
+		consumeAllSpace(l)
```

Comment:
Just call it `ignore` maybe? Its clear that you will be ignoring space from the case `unicode.IsSpace`

Reply:
Calling `ignore` will ignore white-space one at a time, but consuming them all is achieved by `consumeAllSpace`.

Reply:
I still think the `consumeAllSpace` one is still doing the exact same thing so doesn't change anything practically or make things more efficient.